3.13/redundancy_eliminator.md

```diff
@@ -0,0 +1,183 @@
+doing the same in CPython -- by interpreting bytecode instructions.
+The abstract interpreter thus "executes" a bytecode, update its
+symbolic state, and repeats until it reaches a terminator (the end
+of the trace).
+
+### Lattice
+
+**Note: if you are uninterested in the theory behind data-flow analysis,\
+please skip this section.**
+
+The states of information in such an abstract interpretation can be
+represented by a [lattice](https://en.wikipedia.org/wiki/Lattice_(order))
+of values. The lattice itself is an ordered powerset
+of all variables and their associated information.
+
+For example, here is a trivial lattice for the statement 
+```python
+x = 1
+y = 2.0
+```
+:
+
+```mermaid
+flowchart BT
+    A("`Bottom, ie. x: *?*, y: *?*`") --> B("`x: *int*, y: *?*`")
+    A --> C("`x: *?*, y: *float*`")
+    B --> D("`Top, x: *int*, y: *float*`")
+    C --> D
+  
+```
+
+Real Python runtime information
+may contain more complex types, constants, and unions of types.
+
+Each `_Py_UOpsSymType` captures all the information associated
+with a variable at any point in the lattice. For example, it contains
+type information stored as `PyTypeObject *`, and constant values.
+
+For more information regarding lattices in compiler theory,
+Page 27 onwards of this presentation provides a gentle introduction
+https://ilyasergey.net/CS4212/_static/lectures/PLDI-Week-12-dataflow.pdf
+Credits to the courses UPenn Compilers (CIS 341) and NUS Compiler
+Design (CS4212).
+
+The hiearchy of types in the type system can also be represented
+by a lattice. Roughly, it looks like this:
+
+```mermaid
+graph BT;
+    t("True");
+    f("False");
+    bl("bool");
+    nnull(not NULL);
+    nnone(not None);
+    ot(other types);
+    oc(other constants);
+    b("bottom(unknown)");
+    b-->NULL;
+    b-->nnull;
+    nnull-->nnone;
+    nnull-->None;
+    nnone-->ot;
+    nnone-->bl;
+    ot-->oc;
+    oc-->top;
+    None-->top;
+    NULL-->top;
+    bl-->t;
+    bl-->f;
+    t-->top;
+    f-->top;
+```
+
+Note: there should be edges going from each type to `top`, but
+for brevity and clarity we have omitted them.
+
+Edges in the lattice represent a
+[partial order relation](https://en.wikipedia.org/wiki/Partially_ordered_set).
+For example, `bool <: True`, where `<:` is the supertype relation,
+because `bool` is the supertype of `True`. The supertype relation is
+reflexive, antisymmetric, and transitive, and is thus a partial
+order relation.
+
+The [least upper bound](https://www.infinitelymore.xyz/p/lattices)
+of any pair `{a, b}` in the lattice, ie the *join* of `{a, b}`, represents
+the the lowest common subtype of both `a` and `b`. In this case,
+the lowest common subtype of any two types,
+assuming `a != b`, is `top`. This means we can never reach
+`top`, as that is a contradiction -- there is no type
+in CPython that represents the subtype of all types. (Think:
+how can something be a subtype of `int`, `float`, `str`, `bool`, etc.
+all at the same time!)
```

Comment:
I guess that if we find ourselves at _top_, we have projected a trace into unreachable code -- e.g. we traced through something that is guaranteed to raise. Interesting.